keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,330 @@
+
+### Non-Goals
+
+* Tracing in kubernetes controllers
+* Replace existing logging, metrics
+* Change metrics or logging (e.g. to support trace-metric correlation)
+* Access control to tracing backends
+* Add tracing to components outside kubernetes (e.g. etcd client library).
+
+## Proposal
+
+### User Stories
+
+Since this feature is for diagnosing problems with the kubelet, it is targeted at Cluster Operators and Cloud Vendors that manage kubernetes control-planes.
+
+For the following use-cases, I can deploy an OpenTelemetry collector agent as a DaemonSet to collect kubelet trace data from each node's host network. Then, I can deploy a single OpenTelemetry collector to consolidate the kubelet traces. From there, OpenTelemetry trace data can be exported to a tracing backend of choice.  I can use the `EnableOtelTracing` boolean to enable trace exports from kubelet service. I can use the `OpenTelemetryConfig` to configure a trace service name other than the node hostname and the tracing collector port if the OpenTelemetry collector is listening on a different port than the default `0.0.0.0:4317`.
+
+#### Continuous Trace Collection
+
+As a cluster operator or cloud provider, I would like to collect gRPC and HTTP trace data from the transactions between the API server and the kubelet to help debug control-plane problems. Depending on the symptoms I need to debug, I can search span metadata to find a trace which displays the symptoms I am looking to debug.  The sampling rate for trace exports can be configured based on my needs. I can collect each node's kubelet trace data as distinct tracing services to diagnose node issues.
+
+### Tracing Requests and Exporting Spans
+
+We will instrument the kubelet's [gRPC server](https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/kubelet/server/server.go#L190#L192) to intercept gRPC calls and export spans. Also, the http server and http clients will be wrapped with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/v0.21.0/instrumentation/net/http) to generate spans for sampled incoming requests and propagate context with client requests. The gRPC client in the kubelet's Container Runtime Interface (CRI) [Remote Runtime Service](https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/kubelet/cri/remote/remote_runtime.go) and [the CRI streaming package](https://github.com/kubernetes/kubernetes/tree/release-1.21/pkg/kubelet/cri/streaming) will be instrumented to export and propagate trace data. The [Go implementation of OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go) will be used. An [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracegrpc/exporter.go) and an [OTLP trace provider](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/provider.go) along with a [context propagator](https://opentelemetry.io/docs/go/instrumentation/#propagators-and-context) will be configured.
```

Comment:
https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/kubelet/server/server.go#L190#L192 links to the PodResources grpc server, which is used to associate devices with containers.  Did you mean to link to something else?

Reply:
There are a lot of http servers on the kubelet IIRC.  For alpha, i'd prefer to only instrument those for which we have use-cases, and we can expand that list over time.  There are some fairly obscure endpoints (e.g. the pprof endpoint?) which we might not want spans from.

Reply:
not sure where I was going with that (PodResources grpc server), removed it for now, but will instrument kubelet  RemoteRuntimeService, the communication between the kubelet and the node's container runtime. Where is the CRI shim grpc server running (where is that grpc server started?) I'm a bit confused atm on this, but will get straightened out/updated here.   Also, need to gather use-cases

Reply:
resolving this because my original text was wrong/confusing